rpcperms/middleware_handler.go

```diff
@@ -0,0 +1,448 @@
+// dispatches them into the responses channel provided, sending any errors that
+// occur into the error channel provided.
+func (h *MiddlewareHandler) receiveResponses(errChan chan error,
+	responses chan *lnrpc.RPCMiddlewareResponse) {
+
+	for {
+		resp, err := h.receive()
+		if err != nil {
+			errChan <- err
+			return
+		}
+
+		select {
+		case responses <- resp:
+
+		case <-h.done:
+			return
+
+		case <-h.quit:
+			return
+		}
+	}
+}
+
+// sendInterceptRequests handles channel acceptor requests sent to us by our
+// Accept() function, dispatching them to our acceptor stream and coordinating
+// return of responses to their callers.
+func (h *MiddlewareHandler) sendInterceptRequests(errChan chan error,
+	responses chan *lnrpc.RPCMiddlewareResponse) error {
+
+	// Close the done channel to indicate that the interceptor is no longer
+	// listening and any in-progress requests should be terminated.
+	defer close(h.done)
+
+	interceptRequests := make(map[uint64]*interceptRequest)
+
+	for {
+		select {
+		// Consume requests passed to us from our Accept() function and
+		// send them into our stream.
+		case newRequest := <-h.interceptRequests:
+			id := atomic.AddUint64(&h.lastRequestID, 1)
+
+			req := newRequest.request
+			interceptRequests[id] = newRequest
+
+			interceptReq, err := req.ToRPC(id)
+			if err != nil {
+				return err
+			}
+
+			if err := h.send(interceptReq); err != nil {
+				return err
+			}
+
+		// Process newly received responses from our channel acceptor,
+		// looking the original request up in our map of requests and
+		// dispatching the response.
+		case resp := <-responses:
+			requestInfo, ok := interceptRequests[resp.RequestId]
+			if !ok {
+				continue
+			}
+
+			response := &interceptResponse{}
+			switch msg := resp.GetMiddlewareMessage().(type) {
+			case *lnrpc.RPCMiddlewareResponse_Feedback:
+				t := msg.Feedback
+				if t.Error != "" {
+					response.err = fmt.Errorf("%s", t.Error)
+					break
+				}
+
+				// For intercepted responses we also allow the
+				// content itself to be overwritten.
+				if requestInfo.request.Type == TypeResponse &&
+					t.ReplaceResponse {
+
+					response.replace = true
+					protoMsg, err := parseProto(
+						requestInfo.request.ProtoTypeName,
+						t.ReplacementSerialized,
+					)
+
+					if err != nil {
+						response.err = err
+
+						break
+					}
+
+					response.replacement = protoMsg
+				}
+
+			default:
+				return fmt.Errorf("unknown middleware "+
+					"message: %v", msg)
+			}
+
+			select {
+			case requestInfo.response <- response:
+			case <-h.quit:
+			}
+
+			delete(interceptRequests, resp.RequestId)
+
+		// If we failed to receive from our middleware, we exit.
+		case err := <-errChan:
+			log.Errorf("Received an error: %v, shutting down", err)
+			return err
+
+		// Exit if we are shutting down.
+		case <-h.quit:
+			return errShuttingDown
+		}
+	}
+}
+
+type InterceptType uint8
```

Comment:
PR still marked as WIP, but just want to mark here that we're missing `godoc` comments here.